
  
 

##########################################################################
########################################################################## filterImputeData.R
#' Filter genetic variants    
#'
#' @description
#' Filter genetic variants accoring to the imputation quality score with the help of .impute2_info files generated by IMPUTE2. 
#' In this case we keep variants with imputation info score of greater than 0.6.   
#' 
#' @param plink an executable PLINK program in either the current working directory or somewhere in the command path.
#' @param suffix4impute2info the suffix of IMPUTE2 generated files that store the imputation quality score 
#' for each variant, i.e. .impute2_info files.
#' @param impute2infoFile the output file of impute2 info scores which consists of two columns, namely all imputed SNPs and their info scores. 
#' Note that imputed SNPs with more than two alleles are dropped. 
#' @param infoScore  the cutoff of filtering imputation quality score for each variant.   
#' @param badImputeSNPfile the output file of SNPs with bad info scores.  
#' @param prefix4imputedPlink  the prefix of the final imputed PLINK format files. 
#' @param prefix4imputedFilterPlink  the prefix of the final imputed and filtered PLINK format files. 

#' @return  The filtered PLINK format imputed files, a pure text file with all removed SNPs (with bad infoScore) 
#' and a pure text file with the info scores of all imputed SNPs that has two columns: the SNPs and the corresponding info scores.
#' @export 

#' @author Junfang Chen <junfang.chen3@gmail.com> 
#' @examples   
  

## working directory: 6-finalResults
filterImputeData <- function(plink, suffix4impute2info, impute2infoFile, infoScore, badImputeSNPfile, prefix4imputedPlink, prefix4imputedFilterPlink){ 

	# read each .impute2_info file, remove 1st line, add to another file and repeat   
	## get all impute2_info files for each chunk
	files = system(paste0("ls *", suffix4impute2info), intern=TRUE) 
	for (i in 1:length(files)){ 
	 	system(paste0("sed 1d ", files[i], "  >> impute2infoAllvariants.txt")) ## impute2infoAllvariants.txt is the temporal file
	}  
	 
	### only keep SNPs and SNPs with two alleles  
	## impute2infoUpdateTmp.txt > temporal file
	system("grep 'rs' impute2infoAllvariants.txt | awk '{if(length($4)==1 && length($5)==1) print}' | awk '{print $2, $7}' > impute2infoUpdateTmp.txt") 
	system(paste0('mv impute2infoUpdateTmp.txt ', impute2infoFile))
	## added colnames 
	impute2info <- read.table(file=impute2infoFile, stringsAsFactors=F)  
	colnames(impute2info) = c("rs_id", "info") 

	 
	###############  filtering   
	snpWithBadInfo = impute2info[which(impute2info[, "info"] < infoScore), 1]  
	write.table(snpWithBadInfo, file=badImputeSNPfile, quote=FALSE, row.names=FALSE, col.names=FALSE, eol="\r\n", sep=" ")
	### extract filtered SNPs  
	system( paste0(plink, " --bfile ", prefix4imputedPlink, " --exclude ", badImputeSNPfile, " --make-bed --out ", prefix4imputedFilterPlink)) 
	system( 'rm impute2infoAllvariants.txt')

}


